README.md
This is just a placeholder until I come up with something better.


aquatics/CHANGELOG.md
Change log
==========

aquatics uses [Semantic Versioning][1].
The change log is available [on GitHub][2].

[1]: http://semver.org/spec/v2.0.0.html
[2]: https://github.com/githubuser/aquatics/releases

## v0.1.0.0

* Initially created.


aquatics/src/Stuff.hs
module Stuff
where

import Control.Monad
import Data.List

import Prelude
  
roundToStuff :: Int -> Double -> Double
roundToStuff 0 num = (fromIntegral . round $ num * 10) / 10
roundToStuff intExponent num = (fromIntegral . round $ num / f) * f
  where f = 10 ^ (intExponent - 1)

invert :: [] Double -> Double
invert ns = 1 / (sum $ map (1 /) ns)

resistorVals :: Int -> [] Double
-- resistorVals 0 = [ 30000, 56200, 47000, 40200, 20500, 8450, 5600, 5300, 9310 ]
resistorVals intExponent = map (* decade) moreDecades
  where
    firstDecade = [0.1,1,4.7,20,27,33,43,47] -- 0.1,1,4.7,20,27,33,43,47 -- First decade.
    decade = fromIntegral $ 10 ^ intExponent :: Double

moreDecades :: [] Double             
moreDecades =
  [ 47.0    , 100.0   , 250.0   , 500.0   , 1000.0  , 2500.0  , 2700.0
  , 3000.0  , 5000.0  , 5300.0  , 5600.0  , 7150.0  , 7320.0  , 7320.0
  , 7500.0  , 7680.0  , 7870.0  , 7870.0  , 8000.0  , 8200.0  , 8250.0
  , 8450.0  , 8870.0  , 9090.0  , 9100.0  , 9310.0  , 9530.0  , 9760.0
  , 10200.0 , 10500.0 , 10700.0 , 11000.0 , 11300.0 , 20500.0 , 30000.0
  , 40200.0 , 47000.0 , 56200.0
  ]

-- moreDecades =
--   [ 47, 30000, 56200, 47000, 40200, 20500, 8450, 5600
--   , 5300, 9310, 1000, 5000, 9100, 8000, 2500, 2700, 10500
--   , 10700, 7500, 7870, 10200, 7680, 8200, 8250, 11000, 7870
--   , 11300, 3000, 7320, 9530, 9760, 9090, 7150, 8870, 7320
--   , 500, 250, 100
--   ]

-- combos :: [] Double -> [[Double]]
-- combos pp = filter (not . null) $ do
--   ds1 <- ds
--   ds2 <- ds
--   ds3 <- ds
--   ds4 <- ds
--   ds5 <- ds
--   ds6 <- ds -- Anything more than ds6 (e.g. ds7) adds few additional values for tons of compute time.
--   pure $ filter (/= 0) [ds1,ds2,ds3,ds4,ds5,ds6]
--   where ds = 0:pp


caps :: [] Double
caps = map (* (1 / nf)) $ nfs ++ pfs
  where
    nfs = map (* nf) [ 1, 1.2, 1.5, 1.8, 2, 2.2, 3.3, 4.7, 5.6, 6.8, 8.2, 10, 22, 100 ]
    pfs = map (* pf) [ 800, 750, 680, 470, 330, 220, 120, 100, 82 ]
    
    nf = 0.000000001
    pf = 0.000000000001
    









combos :: [] Double -> [[Double]]
combos pp = filter (not . null) $ do
  ds1 <- ds
  ds2 <- ds
  pure $ filter (/= 0) [ds1,ds2]
  where ds = 0:pp

run :: IO ()
run = do
  mapM_ print parts
  print $ length res
  where res = sort $ nub (0 : (map (roundToStuff decade . invert) $ combos (resistorVals decade)))
        parts = map (part res) $ buildRangesByDecade decade
        decade = 0

part :: [] Double -> (Double, Double) -> [] Double
part ds (beg, end) = filter (\n -> n >= beg && n < end) ds

buildRangesByDecade :: Int -> [] (Double, Double)
buildRangesByDecade 0 = zip [0..9] [1 .. 10]
buildRangesByDecade intExponent = zip [decade, decade * 2 ..] [decade * 2, decade * 3 .. decadeSucc]
  where decade     = fromIntegral $ 10 ^ intExponent :: Double
        decadeSucc = fromIntegral $ 10 ^ (intExponent + 1) :: Double
                     
{-
# 0.1,1,4.7,20,27,33,43,47 -- First decade.

time stack exec aquatics-exe
[0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9]
[1.0,1.1,1.2,1.3,1.4,1.5,1.6,1.7,1.8,1.9]
[2.0,2.1,2.2,2.3,2.4,2.5,2.6,2.7,2.8,2.9]
[3.0,3.1,3.2,3.3,3.4,3.5,3.6,3.7,3.8,3.9]
[4.0,4.1,4.2,4.3,4.4,4.5,4.6,4.7,4.8,4.9]
[5.0,5.1,5.2,5.3,5.4,5.5,5.6,5.7,5.8,5.9]
[6.0,6.1,6.2,6.3,6.4,6.5,6.6,6.7,6.8,6.9]
[7.0,7.1,7.2,7.3,7.4,7.5,7.6,7.7,7.8,7.9]
[8.0,8.1,8.2,8.3,8.4,8.5,8.6,8.7,8.8,8.9]
[9.0,9.1,9.2,9.3,9.4,9.5,9.6,9.7,9.8,9.9]
140

real	0m0.377s
user	0m0.525s
sys	0m0.245s

-}


aquatics/src/Resistors.hs
{-# LANGUAGE
    NoImplicitPrelude
  , DeriveDataTypeable
  , DeriveGeneric
  , RankNTypes
  , GeneralizedNewtypeDeriving
  , DerivingStrategies
#-}
-- https://kowainik.github.io/posts/deriving
{-# OPTIONS_GHC -Wno-name-shadowing #-} -- -fwarn-unused-imports
{-|
Module      : Resistors
Copyright   : © 2022 Robert Lee
License     : All Rights Reserved

Maintainers : robert.lee@chicago.vc (Robert)
Stability   : None
Portability : non-portable (GHC extensions)

Contumacy   : Best viewed with unbroken and unwrapped 154 column display.

Description : Provides processing for commissions based on CSV inputs.
Neighborhood: GoldCoast
-}
{-
infixr 9  .
infixr 8  ^, ^^, ⋆⋆
infixl 7  ⋆, /, ‘quot‘, ‘rem‘, ‘div‘, ‘mod‘
infixl 6  +, -
infixr 5  :, ++
infix  4  ==, /=, <, <=, >=, >
infixl 4  <$, <*>, <*, *>, <**>
infixr 3  &&
infixr 2  ||
infixl 1  ?, >>, >>=
infixr 1  =<<, <=<, >=>
infixr 0  $, $!, ‘seq‘
-}

module Resistors
where

-- Explicit Imports

import Data.Char        ( isNumber, toUpper )
import Data.Default     ( Default(..) )
import Data.Data        ( Data, Typeable )
import Data.Hashable
import Data.Ix          ( Ix )
import Data.List        ( sortBy )
import Data.Text        ( Text )
import Data.Word        ( Word8 )
import Data.Bits        ( Bits, FiniteBits )
import GHC.Generics     ( Generic )
import GHC.Num          ( Num )
import Foreign.Storable ( Storable )
import Numeric          ( showFFloat )
import Data.Holmes
import Prelude hiding (String)

-- End of Imports
-- -----------------------------------------------------------------------------------------------------------------------------------------------------

data Geometry a
  = Parallel a [a]
  | Series   a [a]
    deriving (Show, Eq, Ord)

class (Eq a, Show a, Ord a, RealFrac a) => GeometryC a
  where
    seriesToSingle   :: [] a -> a
    parallelToSingle :: [] a -> a
    parallelGeometry :: [] a -> Geometry a
    seriesGeometry   :: [] a -> Geometry a
    round3Geometry   :: Geometry a -> Geometry a
    getScalar  :: Geometry a -> a
    getCount   :: Geometry a -> Int
    isSeries   :: Geometry a -> Bool
    isParallel :: Geometry a -> Bool
    -- defaults
    parallelGeometry ns = Parallel (parallelToSingle ns) ns
    seriesGeometry   ns = Series   (seriesToSingle   ns) ns
    round3Geometry   (Parallel s ns) = Parallel (round3 s) ns
    round3Geometry   (Series   s ns) = Series   (round3 s) ns
    getScalar  (Parallel s _) = s
    getScalar  (Series   s _) = s
    getCount   (Parallel _ n) = length n
    getCount   (Series   _ n) = length n
    isSeries   (Parallel _ n) = False
    isSeries   (Series   _ n) = True
    isParallel g = not $ isSeries g


newtype Seconds = Seconds Double
  deriving stock   ( Data, Generic )
  deriving newtype ( Enum, Eq, Floating, Fractional, Num
                   , Ord, Read, Real, RealFloat, RealFrac
                   )
  deriving newtype (Hashable)

instance Show Seconds where
  show seconds = showFFloat Nothing seconds ""

newtype Ohms = Ohms Double
  deriving stock   ( Data, Generic ) -- Should Show be in stock it will show "Ohms 0", not just numerals.
  deriving newtype ( Enum, Eq, Floating, Fractional, Num
                   , Ord, Read, Real, RealFloat, RealFrac
                   ) -- deriving newtype Show is just numerals (e.g. "0"), not "Ohms 0" etc.
  deriving newtype (Hashable)

instance Show Ohms where
  show ohms = showFFloat Nothing ohms ""

instance GeometryC Ohms where
  seriesToSingle   = sum
  parallelToSingle = recip . sum . map recip
  parallelGeometry ns = Parallel (fromIntegral . round $ parallelToSingle ns) ns -- This could be changed to round only when values are larger than X.

newtype Watts = Watts Double
  deriving stock   ( Data, Generic )
  deriving newtype ( Enum, Eq, Floating, Fractional, Num
                   , Ord, Read, Real, RealFloat, RealFrac
                   )
  deriving newtype (Hashable)

instance Show Watts where
  show watts = showFFloat Nothing watts ""

newtype Tolerance = Tolerance Double
  deriving stock   ( Data, Generic )
  deriving newtype ( Enum, Eq, Floating, Fractional, Num
                   , Ord, Read, Real, RealFloat, RealFrac
                   )
  deriving newtype (Hashable)

instance Show Tolerance where
  show tolerance = showFFloat Nothing tolerance ""

newtype Leads = Leads Word8
  deriving stock   (Generic, Data)
  deriving newtype ( Bits, Bounded, Enum, Eq, FiniteBits, Integral
                   , Ix, Num, Ord, Read, Real, Show, Storable
                   )

newtype Hertz = Hertz Double -- Hz
  deriving stock   ( Data, Generic )
  deriving newtype ( Enum, Eq, Floating, Fractional, Num
                   , Ord , Read, Real, RealFloat, RealFrac
                   )
  deriving newtype (Hashable)

instance Show Hertz where
  show hertz = showFFloat Nothing hertz ""

newtype Duty = Duty Double -- Duty Cycle (a percentage)
  deriving stock   ( Data, Generic )
  deriving newtype ( Enum, Eq, Floating, Fractional, Num
                   , Ord, Read, Real, RealFloat, RealFrac
                   )
  deriving newtype (Hashable)

instance Show Duty where
  show duty = showFFloat Nothing duty ""

-- -----------------------------------------------------------------------------------------------------------------------------------------------------
-- Capacitors

newtype Farads = Farads Double
  deriving stock   ( Data, Generic )
  deriving newtype ( Enum, Eq, Floating, Fractional, Num
                   , Ord, Read, Real, RealFloat, RealFrac
                   )
  deriving newtype (Hashable)

instance Show Farads where
  show farads = showFFloat Nothing farads ""

instance GeometryC Farads where
  seriesToSingle   = recip . sum . map recip
  parallelToSingle = sum

-- -----------------------------------------------------------------------------------------------------------------------------------------------------
-- Inductors

newtype Henrys = Henrys Double
  deriving stock   ( Data, Generic )
  deriving newtype ( Enum, Eq, Floating, Fractional, Num
                   , Ord, Read, Real, RealFloat, RealFrac
                   )
  deriving newtype (Hashable)

instance Show Henrys where
  show henrys = showFFloat Nothing henrys ""

instance GeometryC Henrys where
  seriesToSingle   = sum
  parallelToSingle = recip . sum . map recip

-- -----------------------------------------------------------------------------------------------------------------------------------------------------

data Mounting
  = ThroughHole
  | SMD
  deriving stock (Bounded, Data, Enum, Eq, Generic, Ix, Ord, Read, Show)

data Packaging
  = Radial
  | RadialTubular
  | Axial
  | SR10
  | SR20
  | TO_126
  | TO_220
  | TO_247
  | TO_264
  deriving stock (Bounded, Data, Enum, Eq, Generic, Ix, Ord, Read, Show)

data Resistor =
  Resistor
  { ohms      :: Ohms
  , watts     :: Watts
  , tolerance :: Tolerance
  , leads     :: Leads
  , mounting  :: Mounting
  , packaging :: Packaging
  } deriving stock (Eq, Ord, Read, Show)

instance Default Resistor where
  def = Resistor
        { ohms      = 0.0         -- :: Ohms
        , watts     = 0.0         -- :: Watts
        , tolerance = 0.0         -- :: Tolerance
        , leads     = 2           -- :: Leads
        , mounting  = ThroughHole -- :: Mounting
        , packaging = Axial       -- :: Packaging
        }

-- -----------------------------------------------------------------------------------------------------------------------------------------------------
{-
5v 74LS123
Per the datasheet:
  Keep Rt well under 260kΩ.
  Keep Rt at or above 5kΩ.
  C(pF) has no restrictions.
  Minimum pulse duration 40ns

           K
tw(ns) = 0.28 * Rt(kΩ) * C(pF) * ( 1 + 0.7 / Rt(kΩ) ) -- For non-polarized caps.

example: 10uF cap with 11.8kΩ gives
         0.28 * 11.8 * 10000000 * ( 1 * 0.7 / 11.8 )

round $ twF 0.28 11800 (22 * nf) * 1000000000 -- Gives 77000 nanoseconds.

-- For electrolytic caps. See Figure 2 on pg7 datasheet for Diode placement.
          Kd
tw(ns) = 0.25 * Rt(kΩ) * C(pF) * ( 1 + 0.7 / Rt(kΩ) )
-}

-- | twF(ns) = 0.28 * Rt(kΩ) * C(pF) * ( 1 + 0.7 / Rt(kΩ) ) -- For non-polarized caps.
--   twF(ns) = 0.22 * Rt(kΩ) * C(pF) * ( 1 + 0.7 / Rt(kΩ) ) -- For polarized electrolytics.
twF :: Double -> Ohms -> Farads -> Seconds
twF k ohms farads = convert                           -- Convert from Double to seconds.
                  $ (k * rt * cpF * ( 1 + 0.7 / rt )) -- This expression produces Double in ns.
                  / 1000000000                        -- Convert from ns to seconds.
  where
    rt :: Double
    rt = convert $ ohms / 1000.0

    cpF :: Double
    cpF = convert $ farads * recip pf -- Convert to picofarads from farads.

-- convert :: () =>
convert :: (RealFrac a, RealFrac b) => a -> b
convert = fromRational . toRational

hzDutyToSeconds :: Hertz -> Duty -> Seconds
hzDutyToSeconds hertz duty = dutySeconds
  where
    period = convert $ recip hertz
    dutySeconds = convert $ period * duty

decade :: (RealFrac a) => a -> Int
decade a
  | a < 0 = decade (a * -1)
  | a == 0 = 0
  | a < 0.00000000000000001  = -18  -- atto  (a)
  | a < 0.0000000000000001   = -17
  | a < 0.000000000000001    = -16
  | a < 0.00000000000001     = -15  -- femto (f)
  | a < 0.0000000000001      = -14
  | a < 0.000000000001       = -13
  | a < 0.00000000001        = -12  -- pico  (p)
  | a < 0.0000000001         = -11
  | a < 0.000000001          = -10
  | a < 0.00000001           = -9   -- nano  (n)
  | a < 0.0000001            = -8
  | a < 0.000001             = -7
  | a < 0.00001              = -6   -- micro (µ) -- NB: the µ is unicode not an ascii u
  | a < 0.0001               = -5
  | a < 0.001                = -4
  | a < 0.01                 = -3   -- mili  (m)
  | a < 0.1                  = -2   -- centi (c)
  | a < 1                    = -1   -- deci  (d)
  | a < 10                   = 0
  | a < 100                  = 1    -- deka  (da)
  | a < 1000                 = 2    -- hekto (h)
  | a < 10000                = 3    -- kilo  (k)
  | a < 100000               = 4
  | a < 1000000              = 5
  | a < 10000000             = 6    -- mega  (M)
  | a < 100000000            = 7
  | a < 1000000000           = 8
  | a < 10000000000          = 9    -- giga  (G)
  | a < 100000000000         = 10
  | a < 1000000000000        = 11
  | a < 10000000000000       = 12   -- tera  (T)
  | a < 100000000000000      = 13
  | a < 1000000000000000     = 14
  | a < 10000000000000000    = 15   -- peta  (P)
  | a < 100000000000000000   = 16
  | a < 1000000000000000000  = 17
  | a < 10000000000000000000 = 18   -- exa   (E)
  | otherwise = 999

-- | round3 performs a 3 place precision round.
--   e.g. round3 88940.3 = 88900.0
round3 :: (RealFrac a) => a -> a
round3 a
  | intExponent < 0 = let f = 10 ^ (-1 * intExponent)
                      in (fromIntegral . round $ a * f) / f
  | otherwise = let f = 10 ^ intExponent
                in (fromIntegral . round $ a / f) * f
  where
    intExponent = decade a - 2

µf = Farads 0.000001       -- NB: the µ is unicode not an ascii u
nf = Farads 0.000000001
pf = Farads 0.000000000001

-- | capacitors uses (farads * (1/picofarad)) for rounding. It converts back to farads with (*picofarad).
capacitors :: [] Farads
capacitors = map (* pf) . map (fromIntegral . round) . map (* (recip pf)) $ nfs ++ pfs
  where
    nfs = map (* nf) [ 1, 1.2, 1.5, 1.8, 2, 2.2, 3.3, 4.7, 5.6, 6.8, 8.2, 10, 22, 100 ]
    pfs = map (* pf) [ 800, 750, 680, 470, 330, 220, 120, 100, 82 ]

resistors :: [] Ohms
resistors =
  [ 47.0    , 100.0   , 250.0   , 500.0   , 1000.0  , 2500.0  , 2700.0
  , 3000.0  , 5000.0  , 5300.0  , 5600.0  , 7150.0  , 7320.0  , 7320.0
  , 7500.0  , 7680.0  , 7870.0  , 7870.0  , 8000.0  , 8200.0  , 8250.0
  , 8450.0  , 8870.0  , 9090.0  , 9100.0  , 9310.0  , 9530.0  , 9760.0
  , 10200.0 , 10500.0 , 10700.0 , 11000.0 , 11300.0 , 20500.0 , 30000.0
  , 40200.0 , 47000.0 , 56200.0
  ]

crossProduct :: [] a -> [[a]]
crossProduct pp = do
  a0 <- pp
  a1 <- pp
  pure [a0,a1]

parallelOhmsGeometry :: [] (Geometry Ohms)
parallelOhmsGeometry = map parallelGeometry $ crossProduct resistors

seriesOhmsGeometry :: [] (Geometry Ohms)
seriesOhmsGeometry = map seriesGeometry $ crossProduct resistors

ohmsGeometry :: [] (Geometry Ohms)
ohmsGeometry = parallelOhmsGeometry ++ seriesOhmsGeometry

parallelFaradsGeometry :: [] (Geometry Farads)
parallelFaradsGeometry = map parallelGeometry $ crossProduct capacitors

seriesFaradsGeometry :: [] (Geometry Farads)
seriesFaradsGeometry = map seriesGeometry $ crossProduct capacitors

faradsGeometry :: [] (Geometry Farads)
faradsGeometry = parallelFaradsGeometry ++ seriesFaradsGeometry

-- -----------------------------------------------------------------------------------------------------------------------------------------------------
-- Sample code + WIP

dinesman :: IO (Maybe [ (Text, Defined Int) ]) -- Sample from the Data.Holmes README.
dinesman = do
  let guesses = 5 `from` [1 .. 5]

  mvs <- guesses `satisfying` \[ baker, cooper, fletcher, miller, smith ] -> and'
    [ distinct [ baker, cooper, fletcher, miller, smith ]
    , baker ./= 5
    , cooper ./= 1
    , fletcher ./= 1 .&& fletcher ./= 5
    , miller .> cooper
    , abs' (smith .- fletcher) ./= 1
    , abs' (fletcher .- cooper) ./= 1
    ]
  case mvs of
    Nothing -> pure Nothing
    Just vs -> pure . Just . sortBy (\(_,x) (_,y) -> x `compare` y) $ zip folks vs
  where
    folks = [ "baker", "cooper", "fletcher", "miller", "smith" ]

crossProducts :: [Geometry Ohms] -> [Geometry Farads] -> [] (Geometry Ohms, Geometry Farads)
crossProducts ogs fgs = do
  og <- rangeFiltered
  fg <- fgs
  pure (round3Geometry og, round3Geometry fg)
  where
    rangeFiltered = filter (\geo -> getScalar geo >= 5000 && getScalar geo < 250000) ogs

dutySeconds :: Duty -> Seconds
dutySeconds duty = hzDutyToSeconds 1800 duty

twCalcs :: [] (Geometry Ohms, Geometry Farads) -> [] (Seconds, Geometry Ohms, Geometry Farads)
twCalcs geos = map (\(go,gf) -> (round3 $ twF 0.28 (getScalar go) (getScalar gf), go, gf )) geos


closerP :: Seconds -> Seconds -> Bool
closerP target candidate = candidate >= lower && candidate <= upper
  where
    fudge = target * 0.01 -- 2%
    upper = target + fudge
    lower = target - fudge

componentCount :: (GeometryC a, GeometryC b) => Geometry a -> Geometry b -> Int
componentCount a b = getCount a + getCount b

-- map (\(s,_,_) -> s) . filter (\(s,_,_) -> (round3 $ dutySeconds 0.053) == s) $  twCalcs $ crossProducts ohmsGeometry faradsGeometry
-- take 1 . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,o,_) -> isSeries o && (round3 $ dutySeconds 0.053) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry
-- length . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,o,_) -> isSeries o && (round3 $ dutySeconds 0.053) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry
-- length . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,o,_) -> isParallel o && (round3 $ dutySeconds 0.053) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry
-- length . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,_,f) -> isSeries f && (round3 $ dutySeconds 0.053) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry
-- length . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,_,f) -> isParallel f && (round3 $ dutySeconds 0.053) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry
-- length . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,o,f) -> isSeries o && isParallel f && (round3 $ dutySeconds 0.053) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry
-- length . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,o,f) -> isSeries f && isParallel o && (round3 $ dutySeconds 0.053) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry

-- length . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,o,f) -> isSeries o && isParallel f && (round3 $ dutySeconds 0.076) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry
-- length . map (\(_,o,f) -> (componentCount o f, o, f)) . filter (\(s,o,f) -> isSeries o && isParallel f && (round3 $ dutySeconds 0.06) == s) $ twCalcs $ crossProducts ohmsGeometry faradsGeometry

fnoodles :: [] Duty -> [] Int
fnoodles ds
  = map mapF ds
  where
    mapF :: Duty -> Int
    mapF d
      = length
      . map (\(_,o,f) -> (componentCount o f, o, f))
      $ filter (\(s,o,f) -> isSeries o && isParallel f && (round3 $ dutySeconds d) == s)
        calcs

{-# NOINLINE calcs #-}
calcs :: [] (Seconds, Geometry Ohms, Geometry Farads)
calcs = twCalcs $ crossProducts ohmsGeometry faradsGeometry

targets :: [] Duty
targets = map (/ 100.0)
  [ 5.3
  , 5.4
  , 5.5
  , 5.6
  , 5.7
  , 5.8
  , 5.9
  , 6.0
  , 6.1
  , 6.2
  , 6.3
  , 6.4
  , 6.5
  , 6.6
  , 6.8
  , 7.6
  ]
